website/source/docs/auth/aws-iam.html.md

```diff
@@ -0,0 +1,606 @@
+---
+layout: "docs"
+page_title: "Auth Backend: AWS-IAM"
+sidebar_current: "docs-auth-aws-iam"
+description: |-
+  The aws-iam backend allows automated authentication AWS IAM principals.
+---
+
+# Auth Backend: aws-iam
+
+The aws-iam auth backend provides a secure introduction mechanism for AWS IAM
+principals. This allows you to reduce the problem of securely introducing a
+Vault token to the problem of securely introducing AWS IAM credentials, which
+AWS has already solved in a number of use cases, such as via EC2 Instance
+Profiles and IAM roles attached to Lambda functions. It also allows you to have
+a consistent workflow between developers working on a local laptop with tools
+such as Hologram.
+
+## Comparison with AWS-EC2 Authentication Backend
+
+The AWS-IAM and AWS-EC2 authentication backends serve similar purposes. Both
```

Comment:
Just throwing this out there as a possible future enhancement....

When you make this authentication call from an EC2 instance inside an IAM instance profile, Vault will actually see the instance ID of the calling instance -- Vault sees `arn:aws:iam::0123456789012:role/Role/InstanceId`. If Role's AssumeRolePolicyDocument has `"Service": "ec2.amazonaws.com"` as the only `Principal`, then you have pretty good assurance that `InstanceId` is in fact the instance ID of the EC2 instance that is authenticating to Vault. (It wouldn't give you region, but the chances of two instances in different regions in the same account having the exact same instance ID seems pretty small to me, and I think there are easy ways around that.) Vault could extract out the instance ID, and then that would allow you to do pretty much all of the same kinds of filtering we're talking about with aws-ec2: AMI ID, VPC ID, Subnet ID, tags, etc.

I'm **not** suggesting implementing this today (or even in this release), and I think the code might get a little messy, but since you're both engaged on this topic now, I thought I'd throw this idea out there to see if you think it might be something worth pursuing in the future, or if you think any marginal benefit over what's already provided by aws-ec2 just wouldn't be worth it.